Authentication

poncem91 opened the pull request:
Comment form doesn't load unless user has logged in. If user isn't logged in, they are provided a link to log in and then redirected to the page where they'll now be able to see the comment box to submit. The email field was removed from the comment form since now this can be retrieved directly from the login credentials.
Delete All Comments button has been changed to Delete All Your Comments, so when clicked it will only delete all the comments the current logged user has previously submitted. 
Similarly, the delete X button to delete single comments will only show for comments submitted by the logged in user.
In the servlet also only comments matching the currently logged userId can be deleted.

poncem91 committed:
add basic authentication implementation

Comment form doesn't load unless user has logged in. If user isn't logged in, they are provided a link to log in and then redirected to the page where they'll now be able to see the comment box to submit. The email field was removed from the comment form since now this can be retrieved directly from the login credentials. A logout link is provided as well.

poncem91 committed:
fix redirects

poncem91 committed:
change doDelete single comment to restrict who can delete comments

Inside the single comment delete functionality, the servlet checks to see if the userId of the logged on user matches the one of the comment that needs to be deleted. It will only allow single comments to be deleted by those who wrote them.

poncem91 committed:
change front-end comment history based on user login

Comments that have been written by users logged in will have the X delete button, comments submitted by other users will not show that delete button. The delete all comments button has been changed to only delete all of the logged user's comments. This button will only appear if a user is logged in.

poncem91 commented:
I stored the logged-in user's userId as a data- attribute in the contactme `<div>` so that it can later be referenced in the JS file, this doesn't feel all that right, would it be better to store this in the `<body>` or approach this differently?

poncem91 commented:
I'm also getting the following console error message when deleting any comments:
`DELETE https://8080-dfe2475d-f262-4cff-8707-9112c4571804.us-west1.cloudshell.dev/ 405 (Method Not Allowed)`
Everything still works regardless, but I can't figure out why the error is happening. Tech support seemed to think it might have to do with the Users API but couldn't find an answer. 
Should I try moving this all to a different servlet to avoid using the DELETE method?

peap commented:
Apparently I have to leave replies to your comments down here in a new comment thread :/. Boo, GitHub.

> I stored the logged-in user's userId as a data- attribute in the contactme `<div>` so that it can later be referenced in the JS file, this doesn't feel all that right, would it be better to store this in the `<body>` or approach this differently?

Typically, a derivative of the user's credentials are stored in a special header or a cookie, which the browser subsequently sends with every request in a HTTP header; some frameworks call this an "authentication token". Then the serving framework can load the token from the header and map it back to a user object, which you'd have access to in your servlet handlers via the `request` object.

For now, I think it's fine to attach the username wherever. Since it's not particular to that `div`, adding an attribute something closer to the page root sounds good to me: the `body` or even the `html` element would be fine.

> I'm also getting the following console error message when deleting any comments:
> `DELETE https://8080-dfe2475d-f262-4cff-8707-9112c4571804.us-west1.cloudshell.dev/ 405 (Method Not Allowed)`
> Everything still works regardless, but I can't figure out why the error is happening. Tech support seemed to think it might have to do with the Users API but couldn't find an answer.
> Should I try moving this all to a different servlet to avoid using the DELETE method?

That looks to me like your servlet handling the "/" URL doesn't implement the HTTP DELETE method, i.e., has no `doDelete` method in the class. (In fact, there's no custom servlet for "/", at all.) Watching the network tab in my browser, I do see it doing DELETE calls both to /comments and /. So then the question is, what's making that request? 

I believe this is a result of the redirect response being returned by your DELETE handler for /comments; `fetch`'s default behavior is to follow redirects (https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch), so it issues a second fetch request as part of the response handling. (Apparently it does this using the same HTTP method as the original request.) You can disable that behavior by setting the `redirect` option in your request, or by having your servlet return something else. Probably better to just have your servlet respond with a simple 200 OK; there's no need for it to redirect, IMO.

peap commented on portfolio/src/main/java/com/google/sps/data/Comment.java:
```diff
@@ -24,13 +24,15 @@
   private final String email;
```
Does `userId` make `email` redundant? If so, can we get rid of it?

poncem91 replied:
The only thing I still need the email for is for generating a link to the email of the user who posted a comment in the comments history, but perhaps for privacy reasons it might not be the greatest idea to include this now that they're not given a choice of whether or not to include their email... I'll change this!

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -72,8 +74,9 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       String email = (String) entity.getProperty("email");
       String message = (String) entity.getProperty("message");
       long timestamp = (long) entity.getProperty("timestamp");
+      String userId =  (String) entity.getProperty("userId");
```
The other advantage of using authentication tokens is that someone intercepting their POST requests to your site can't (as easily) imitate another user. If you trust the user's browser to provide their userId, users can trivially replace that field with a user ID of their choice.

But this is fine, for now :)

poncem91 replied:
Lets hope no hackers try to mess with my site anytime soon :) 
I'll read up on authentication tokens still, I'm curious how the security aspect of that works

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -100,27 +106,35 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     
-    // Redirect back to the HTML page.
-    response.sendRedirect("/index.html");
+    // Redirect back to the jsp page.
```
You've remembered to update your comment this time, but comments like these that talk about specific implementation details (e.g., the language chosen to write the page at your redirect destination) tend to get stale. Better would be to say something more general about the intentions of the code, e.g., "Redirect back to the home page".

poncem91 replied:
This makes perfect sense, will change this, thanks!!

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -100,27 +106,35 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
         for (Entity entity : results.asIterable()) {
-            datastore.delete(entity.getKey());
+            String commentUserId = (String) entity.getProperty("userId");
+            if (userId.equals(commentUserId)){
```
Alternatively, let the database do the work by filtering the original query down to the comments owned by the logged-in user.

poncem91 replied:
That's smart, I'll do that. Thanks!!

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -135,18 +149,27 @@ public void doDelete(HttpServletRequest request, HttpServletResponse response) t
+            
+
+            // Only allows comments written by user logged in to be deleted
+            if (userId.equals(commentUserId)){
```
Likewise, we can make datastore do this work by handing it two filters with `CompositeFilter`: https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/Query.CompositeFilter?hl=en#class-query.compositefilter.

poncem91 replied:
Thanks! I hadn't read up on `CompositeFilter` before. This'll work better!

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -135,18 +149,27 @@ public void doDelete(HttpServletRequest request, HttpServletResponse response) t
 
+    // Redirect back to the jsp page.
+    response.getWriter().println();
+    response.sendRedirect("/");
```
This is why the browser is requesting your root URL with the DELETE method and getting a 405 error; `fetch` automatically follows this redirect after getting the response from its DELETE /comments request.

poncem91 replied:
Ahhh, that makes sense. I'll remove the redirect to prevent this error. Thanks for troubleshooting this, I appreciate it!

peap commented on portfolio/src/main/webapp/index.jsp:
```diff
@@ -128,32 +128,45 @@ <h1>Work Experience</h1>
       <!-- Contact Me Section -->
-      <section id="contactme">
+
+      <%@ page import = "com.google.appengine.api.users.UserService, com.google.appengine.api.users.UserServiceFactory" %>
```
Unless you saw differently in a style guide, I think these imports should each get their own directive:

```
<%@ page import = "com...UserService" %>
<%@ page import = "com...UserServiceFactory" %>
```

poncem91 replied:
I think I saw an example on stackoverflow, but I can't speak for the legitimacy of it. I'll change this to two lines.

peap commented on portfolio/src/main/webapp/index.jsp:
```diff
@@ -128,32 +128,45 @@ <h1>Work Experience</h1>
+            String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+
+          %>
+          <p class="login-messages">(You are logged in as <%=userService.getCurrentUser().getEmail()%>, to logout <a href="<%=logoutUrl%>">click here</a>)</p>
```
"click here" isn't generally what you want in the text of a hyperlink, since, e.g., screen readers end up saying "click here" to their users without much context when they tab through the page looking for links. Rather, make the link text descriptive and optionally trust that users know they can click on it, as you did for the log-in link.

poncem91 replied:
Got it! To clarify, would it be best to rephrase also the "Click here to login" to just "Login" (where the whole word is the link)?

peap replied:
Either sounds fine to me.

peap reviewed (changes requested):
Yay, JSP!

poncem91 commented:
JSP made things so much simpler, thanks for suggesting me to use that!
Since Github won't let me reply directly to your comment addressing the userId and the error message:
I'll change the userId data attribute to the `<html>` tag and hope no one messes with my site for now. Maybe in the future I can make those changes. 
Ah, I thought the "/" redirected would automatically go to "/index.jsp" since it seemed to load directly there when I run the devserver. I'll remove the redirect altogether though since I don't think it's necessary.
Thanks!!

poncem91 committed:
change doDelete implementation, move data-user-id, fix other issues

The doDelete servlet method implementation was changed to make it more efficient by filtering the queries before preparing the query results, and the redirect to home was removed to fix the Error 405. The data- attribute that stores the user's id was moved to the body element. The email link previously included in user's past comments has been removed, since there is no longer an option to opt out of including you email. The login and logout wording above the comment history has been changed so that the text inside the hyperlink is exactly the function of the link.
